resources/timerbox.ts

```diff
@@ -1,111 +1,141 @@
 class TimerBox extends HTMLElement {
-  static get observedAttributes() {
+  _start: number;
+  _duration: number;
+  rootElement: HTMLDivElement;
+  largeBoxElement: HTMLDivElement;
+  largeBoxBackgroundElement: HTMLDivElement;
+  largeBoxForegroundElement: HTMLDivElement;
+  smallBoxElement: HTMLDivElement;
+  smallBoxBackgroundElement: HTMLDivElement;
+  smallBoxForegroundElement: HTMLDivElement;
+  timerElement: HTMLDivElement;
+  kLargeSize: number;
+  kSmallSize: number;
+  kBorderSize: number;
+  kFontSize: number;
+  kAnimateMS: number;
+  _threshold: number;
+  _bg: string;
+  _fg: string;
+  _scale: number;
+  _valueScale: number;
+  _towardTop: boolean;
+  _styleFill: boolean;
+  _hideAfter: number;
+  _bigAtZero: boolean;
+  _roundUpThreshold: boolean;
+  _connected: boolean;
+  _hideTimer: number | null;
+  _timer: number | null;
+  _animationFrame: number | null;
+
+  static get observedAttributes(): string[] {
     return ['duration', 'threshold', 'bg', 'fg', 'toward', 'style', 'hideafter', 'bigatzero', 'roundupthreshold'];
   }
 
   // The full duration of the current countdown. When this is changed,
   // the countdown restarts at the new value. If set to 0 then countdowns
   // are stopped.
-  set duration(d) {
-    this.setAttribute('duration', d);
+  set duration(d: string | null) {
+    this.setAttribute('duration', d ?? '');
   }
-  get duration() {
+  get duration(): string | null {
     return this.getAttribute('duration');
   }
 
   // Below this a large box is shown, above it a small box is shown.
-  set threshold(t) {
-    this.setAttribute('threshold', t);
+  set threshold(t: string | null) {
+    this.setAttribute('threshold', t ?? '');
   }
-  get threshold() {
+  get threshold(): string | null {
     return this.getAttribute('threshold');
   }
 
   // All visual dimensions are scaled by this.
-  set scale(s) {
-    this.setAttribute('scale', s);
+  set scale(s: string | null) {
+    this.setAttribute('scale', s ?? '');
   }
-  get scale() {
+  get scale(): string | null {
     return this.getAttribute('scale');
   }
 
   // The displayed value is scaled by this.
-  set valuescale(s) {
-    this.setAttribute('valuescale', s);
+  set valuescale(s: string | null) {
+    this.setAttribute('valuescale', s ?? '');
   }
-  get valuescale() {
+  get valuescale(): string | null {
```

Comment:
Is it possible for any of these to return null?

Reply:
`this.getAttribute` returns `string | null`, or we can make a null-check here to ensure it returns `string` only?

Reply:
Sorry, to match #2649 (as per the discussion in https://github.com/quisquous/cactbot/pull/2532#discussion_r602517869), I think these should all be `string | null` after all, here and in the other resource PRs.

Reply:
Done.